In [31]:
#import dependencies
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler


In [2]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))

In [ ]:
# Convert categorical data to numeric and separate target feature for training data


In [29]:
dum_train = pd.get_dummies(train_df)
dum_train

Unnamed: 0   index  loan_amnt  int_rate  installment  annual_inc  \
0           57107   57107    13375.0    0.1797       483.34    223000.0   
1          141451  141451    21000.0    0.1308       478.68    123000.0   
2          321143  321143    20000.0    0.1240       448.95    197000.0   
3           11778   11778     3000.0    0.1240       100.22     45000.0   
4          169382  169382    30000.0    0.1612      1056.49    133000.0   
...           ...     ...        ...       ...          ...         ...   
12175      354912  354912    19975.0    0.2565       801.09     28000.0   
12176      354944  354944    15000.0    0.1774       540.34     50000.0   
12177      354973  354973     3600.0    0.1862       131.28     60000.0   
12178      355002  355002    15000.0    0.0881       475.68     62000.0   
12179      355307  355307    15000.0    0.1774       540.34     54080.0   

         dti  delinq_2yrs  inq_last_6mths  open_acc  ...  \
0      29.99          0.0             0.0      15.0  ...   
1      11.26          2.0             0.0      16.0  ...   
2      11.28          0.0             0.0      12.0  ...   
3      18.08          0.0             0.0      12.0  ...   
4      27.77          0.0             2.0      13.0  ...   
...      ...          ...             ...       ...  ...   
12175  28.42          0.0             0.0      15.0  ...   
12176  23.43          4.0             0.0      16.0  ...   
12177  28.80          0.0             1.0      14.0  ...   
12178  11.44          0.0             0.0       5.0  ...   
12179  21.66          1.0             1.0       5.0  ...   

       loan_status_low_risk  pymnt_plan_n  initial_list_status_f  \
0                         1             1                      0   
1                         1             1                      0   
2                         1             1                      0   
3                         1             1                      0   
4                         1             1                      0   
...                     ...           ...                    ...   
12175                     0             1                      0   
12176                     0             1                      0   
12177                     0             1                      0   
12178                     0             1                      0   
12179                     0             1                      1   

       initial_list_status_w  application_type_Individual  \
0                          1                            1   
1                          1                            1   
2                          1                            1   
3                          1                            1   
4                          1                            1   
...                      ...                          ...   
12175                      1                            1   
12176                      1                            1   
12177                      1                            1   
12178                      1                            0   
12179                      0                            1   

       application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                               0                1                0   
1                               0                1                0   
2                               0                1                0   
3                               0                1                0   
4                               0                1                0   
...                           ...              ...              ...   
12175                           0                1                0   
12176                           0                1                0   
12177                           0                1                0   
12178                           1                1                0   
12179                           0                1    

In [7]:
# Convert categorical data to numeric and separate target feature for testing data
dum_test = pd.get_dummies(test_df)
dum_test.head()

Unnamed: 0  index  loan_amnt  int_rate  installment  annual_inc    dti  \
0       67991  67991    40000.0    0.0819       814.70    140000.0  19.75   
1       25429  25429     6000.0    0.1524       208.70     55000.0  11.52   
2       38496  38496     3600.0    0.1695       128.27     42000.0   6.74   
3       19667  19667    20000.0    0.1524       478.33    100000.0  12.13   
4       37505  37505     3600.0    0.1240       120.27     50000.0  16.08   

   delinq_2yrs  inq_last_6mths  open_acc  ...  loan_status_high_risk  \
0          0.0             1.0      18.0  ...                      0   
1          2.0             0.0       8.0  ...                      0   
2          0.0             0.0       6.0  ...                      0   
3          0.0             2.0       7.0  ...                      0   
4          0.0             3.0       6.0  ...                      0   

   loan_status_low_risk  pymnt_plan_n  initial_list_status_f  \
0                     1             1                      0   
1                     1             1                      0   
2                     1             1                      0   
3                     1             1                      0   
4                     1             1                      0   

   initial_list_status_w  application_type_Individual  \
0                      1                            1   
1                      1                            1   
2                      1                            1   
3                      1                            1   
4                      1                            1   

   application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                           0                1                0   
1                           0                1                0   
2                           0                1                0   
3                           0                1                0   
4                           0                1                0   

   debt_settlement_flag_N  
0                       1  
1                       1  
2                       1  
3                       1  
4                       1  

[5 rows x 95 columns]

In [21]:
dum_test.head(50)

Unnamed: 0  index  loan_amnt  int_rate  installment  annual_inc    dti  \
0        67991  67991    40000.0    0.0819       814.70    140000.0  19.75   
1        25429  25429     6000.0    0.1524       208.70     55000.0  11.52   
2        38496  38496     3600.0    0.1695       128.27     42000.0   6.74   
3        19667  19667    20000.0    0.1524       478.33    100000.0  12.13   
4        37505  37505     3600.0    0.1240       120.27     50000.0  16.08   
5        26517  26517    27000.0    0.0819       549.93     57000.0  31.41   
6        30063  30063    15000.0    0.0881       310.00     50000.0  24.44   
7        31659  31659    19000.0    0.0881       602.52     70000.0  30.94   
8        15003  15003    27575.0    0.1171       912.07    142000.0  17.90   
9        75464  75464    31500.0    0.1774       795.45    100000.0  25.05   
10       37768  37768    26400.0    0.1033       565.22     70000.0  24.48   
11       35473  35473     5000.0    0.1308       168.67     45370.0  28.70   
12       16700  16700    24000.0    0.0756       747.22    320000.0  17.99   
13       51646  51646     3600.0    0.0881       114.17    138000.0  19.98   
14       48811  48811    12000.0    0.1862       308.79     45000.0  33.17   
15       16394  16394    12000.0    0.1308       273.53     85000.0  26.36   
16       33289  33289    28000.0    0.0756       871.75     67000.0  37.70   
17       35508  35508    10425.0    0.2880       435.74    250000.0  19.53   
18        5635   5635     9000.0    0.0819       183.31     60000.0  12.58   
19       16298  16298    20000.0    0.1033       428.20    109000.0  26.69   
20       60751  60751    12750.0    0.1612       449.01     96000.0  24.93   
21       23360  23360     2000.0    0.0756        62.27     17300.0   6.94   
22       53839  53839    23150.0    0.0756       720.75    130500.0  22.61   
23       22016  22016     4625.0    0.1774       166.61     32000.0  28.02   
24       75140  75140    20400.0    0.0756       635.13     45000.0  19.57   
25       66157  66157    15000.0    0.2055       402.02     39500.0  15.41   
26       59944  59944    17000.0    0.0881       351.33     58000.0  35.30   
27       57242  57242    16000.0    0.1524       382.66     80000.0  31.71   
28       17413  17413    21000.0    0.1430       491.91     86300.0   4.96   
29       77265  77265     8000.0    0.1240       267.25     36400.0  26.01   
30       14580  14580     9000.0    0.1033       291.81     72000.0  38.93   
31       21818  21818     8400.0    0.0646       257.30     57100.0  43.40   
32       63571  63571    10000.0    0.0819       314.25     42000.0  14.80   
33       30310  30310     5000.0    0.1774       180.12     35000.0   6.76   
34       19385  19385     8000.0    0.1430       274.59    110000.0   9.76   
35       73076  73076    10000.0    0.1308       337.33     75000.0  13.89   
36       60188  60188    15950.0    0.1308       538.04     58000.0  29.59   
37       62630  62630    10000.0    0.1102       217.53     62500.0  13.94   
38       40653  40653     6375.0    0.2055       238.71     59000.0  24.69   
39       74059  74059    15000.0    0.1612       528.25     60000.0  20.58   
40       58499  58499     2000.0    0.2305        77.48     50000.0  19.47   
41       71968  71968    18000.0    0.0881       570.81     60000.0  24.06   
42       71664  71664     5600.0    0.0702       172.97     71000.0  17.68   
43       44688  44688    21000.0    0.2055       786.34    139000.0  17.11   
44       27876  27876     2500.0    0.0881        79.28    100000.0   4.88   
45       67454  67454    20000.0    0.1102       435.05     98000.0  30.00   
46       68213  68213     5300.0    0.0819       166.55     27000.0  16.98   
47       10755  10755    15000.0    0.0646       459.47     98000.0  19.90   
48         567    567    19125.0    0.0819       600.99     95800.0  21.07   
49       55847  55847    16000.0    0.1430       549.18    105200.0  16.08   

    delinq_2yrs  inq_last

In [9]:
#view the columns in the training data
dum_train.columns

Index(['Unnamed: 0', 'index', 'loan_amnt', 'int_rate', 'installment',
       'annual_inc', 'dti', 'delinq_2yrs', 'inq_last_6mths', 'open_acc',
       'pub_rec', 'revol_bal', 'total_acc', 'out_prncp', 'out_prncp_inv',
       'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc'

In [14]:
#view the columns in the testing data to check for differences
dum_test.columns

Index(['Unnamed: 0', 'index', 'loan_amnt', 'int_rate', 'installment',
       'annual_inc', 'dti', 'delinq_2yrs', 'inq_last_6mths', 'open_acc',
       'pub_rec', 'revol_bal', 'total_acc', 'out_prncp', 'out_prncp_inv',
       'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc'

In [17]:
#check the number of columns to confirm that there is only one difference - training data has 'debt_settlement_flag_Y' while test data does not
print(len(dum_train.columns))
print(len(dum_test.columns))

96
95


In [ ]:
# add missing dummy variables to testing set

### Since the column missing in the test data would not contribute to the training of the model in a meaningful way and the 'debt_settlement_flag_N' column provides the information (any row would be flagged Y or N so we can extrapolate from the N column), the 'debt_settlement_flag_Y' column can be dropped from the training data instead of adding a dummy column to the test data.

In [20]:
#drop the 'debt_settlement_flag_Y' column from the training data to resolve the difference between training and test data
dum_train_drop = dum_train.drop('debt_settlement_flag_Y', axis=1)

In [24]:
# separate the target from the features for the training data and drop columns from X that would contaminate the training data
X_train = dum_train_drop.drop(['loan_status_high_risk', 'loan_status_low_risk'], axis=1)
y_train = dum_train_drop['loan_status_high_risk']

In [26]:
# separate the target from the features for the test data and drop columns from X that would contaminate the test data
X_test = dum_test.drop(['loan_status_high_risk', 'loan_status_low_risk'], axis=1)
y_test = dum_test['loan_status_high_risk']

### For this data, I believe that a Random Forest model will perform better than a Logistic Regression model.

In [28]:
# Train the Logistic Regression model on the unscaled data and print the model score
LR = LogisticRegression()
LR.fit(X_train, y_train)
print(f"Logistic Regression Testing Data Score: {LR.score(X_test, y_test)}")

Logistic Regression Testing Data Score: 0.5253083794130158


/opt/anaconda3/envs/dev/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [30]:
# Train a Random Forest Classifier model and print the model score
RF = RandomForestClassifier(random_state=1, n_estimators=500)
RF.fit(X_train, y_train)
print(f'Random Forest Testing Data Score: {RF.score(X_test, y_test)}')

Random Forest Testing Data Score: 0.6135686941726924


### The Random Forest model did perform better than the Logistic Regression model. However, neither did well and improvements need to be made. Since the Random Forest model did better to begin with, I would choose to work on improving that one and abandon the Logistic Regression model at this stage. 

In [32]:
# Scale the data
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)


### Now that the data has been scaled, the performance of the models may change. However, I think that the Random Forest model is still likely to outperform the Logistic Regression model.

In [35]:
# Train the Logistic Regression model on the scaled data and print the model score
LR2 = LogisticRegression()
LR2.fit(X_train_scaled, y_train)
print(f"Logistic Regression Testing Data Score: {LR2.score(X_test_scaled, y_test)}")


Logistic Regression Testing Data Score: 0.7218205019140791


/opt/anaconda3/envs/dev/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [42]:
# Train a Random Forest Classifier model on the scaled data and print the model score

RF2 = RandomForestClassifier(random_state=1, n_estimators=500)
RF2.fit(X_train_scaled, y_train)
print(f'Random Forest Testing Data Score: {RF2.score(X_test_scaled, y_test)}')

Random Forest Testing Data Score: 0.613143343258188


### Scaling the data did not improve the Random Forest model's score; in fact, it scored slightly lower on the scaled data. I was wrong in my prediction that the Random Forest model would still perform better, than the Logistic Regression model this time. The Logistic Regression model's score improved significantly, from about 0.525 to about 0.722. This would seem to indicate that scaling the data is very important when working with a Logistic Regression model and less so when working with a Random Forest model. The performance of both models is not particularly impressive and in order for either to be useful modifications would need to be made to improve them significantly.